### PR TITLE
Global Styles: Heading element UI controls

### LINKED ISSUE #41974: Global Styles: Heading element UI controls
## What problem does this address?
Heading elements cannot be edited using the Global Styles UI.

## What is your proposed solution?
It would be good to add heading elements to the list of elements that can be customized in the Global Styles UI:

Screenshot 2022-06-27 at 13 15 32

In addition, it would be good to have more granular control over heading (h1, h2, h3, h4, h5, h6) elements. Right now it's impossible to set rules for these elements via the UI, it can only be done via theme.json.

There is an older issue for when this was possible at a block level: https://github.com/WordPress/gutenberg/issues/31695

### PR DESCRIPTION
**Pull request by carolinan**
## What?

 Add heading elements to the list of elements that can be customized in the Global Styles UI.
 Fixes https://github.com/WordPress/gutenberg/issues/41974
 
 ## Why?

To allow customizing heading elements and not only heading blocks.

## How?

Adds a new heading color screen to the global styles sidebar.

## Screenshots or screencast 
Updated August 1:
The label above the buttons with the heading level selection now says "Select heading level".
The heading above the text color and background color options indicates which heading the style will be applied to.
The text over the color option says "Text color for all heading levels"

With H2 selected:
The text above the color option says "text color for H2"

https://user-images.githubusercontent.com/7422055/182293289-9fbf441b-2550-4093-a2c0-29fdc6901264.mov

### CODE REVIEW

**Comment by carolinan** [comment]
I'd like to know if this design is acceptable, and then I can continue with the typography selection screen.

**Comment by scruffian** [comment]
Thanks for getting started on this @carolinan. The design looks good to me. We can always iterate on it later if we want to, so I wouldn't let that stop you proceeding.

**Comment by carolinan** [comment]
Should size and line height be available for the "all headings" option? Heading levels are intended to have different sizes.

**Comment by carolinan** [comment]
I don't think this is correct:
- On the "root" level (the first screen)  Typography > Headings, and Color > Headings is available
- On block level, Color > Headings is available.

What is the expected result?

**Comment by scruffian** [comment]
> Should size and line height be available for the "all headings" option? Heading levels are intended to have different sizes.

Line height doesn't seem to be a problem, but I don't think size should be there.

**Comment by scruffian** [comment]
> On the "root" level (the first screen) Typography > Headings and Color > Headings is available
> On block level, Color > Headings is available.

I think this is correct - it should be possible to set the heading element globally, and for individual blocks.

This PR is looking great. Is there much more to do?

**Comment by carolinan** [comment]
I hid the font size option for the "all headings" option. I wonder if it should also be hidden for links?
It is ready for review once tests pass.

File: packages/edit-site/src/components/global-styles/screen-heading-color.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		'color.customGradient',
+		name
+	);
+	const [ isTextEnabled ] = useSetting( 'color.text', name );
+	const [ isBackgroundEnabled ] = useSetting( 'color.background', name );
+
+	const colorsPerOrigin = useColorsPerOrigin( name );
+	const gradientsPerOrigin = useGradientsPerOrigin( name );
+
+	const hasTextColor =
```
Comment (scruffian) [inline packages/edit-site/src/components/global-styles/screen-heading-color.js:41]: All of these supports (`hasTextColor`, `hasBackgroundColor`, `hasGradientColor`) are quite verbose, and we probably have this same logic elsewhere. Can we extract this to some kind of hook or shared function? That could be a follow-up PR....

File: packages/edit-site/src/components/global-styles/screen-heading-color.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+				<ToggleGroupControl
+					label={ __( 'Select heading level' ) }
+					hideLabelFromVision={ true }
+					value={ selectedLevel }
+					onChange={ setCurrentTab }
+					isBlock
+				>
+					<ToggleGroupControlOption
+						value="heading"
+						label={ __( 'All' ) }
```
Comment (scruffian) [inline packages/edit-site/src/components/global-styles/screen-heading-color.js:137]: Does this need a translators comment?

**Review by scruffian — APPROVED** [review]
Great work, thank you.

File: packages/edit-site/src/components/global-styles/screen-heading-color.js (hunk 1 above)
Comment (carolinan) [reply packages/edit-site/src/components/global-styles/screen-heading-color.js:41]: Yes I fully agree.

File: packages/edit-site/src/components/global-styles/screen-heading-color.js (hunk 2 above)
Comment (carolinan) [reply packages/edit-site/src/components/global-styles/screen-heading-color.js:137]: Added.

**Comment by youknowriad** [comment]
Just stumbled upon this PR. How do we decide which context: (global, block, element...) supports "heading colors"?

It seems that right now this button randomly shows up for most blocks that support colors? Do we want opt-in/opt-out...

**Comment by carolinan** [comment]
We don't decide, that's what I was trying to get to with this comment, but I was too vague:

> I don't think this is correct:
> 
> * On the "root" level (the first screen)  Typography > Headings, and Color > Headings is available
> * On block level, Color > Headings is available.
> 
> What is the expected result?

I was first thinking it could be opt-in or an allowed list, but would that not affect third party blocks negatively?
So it probably needs to be opt-out now since some time has passed.

For example, the verse block has a heading color option but it can't have headings.
Removing the heading color from the options should not cause any issues since the option doesn't change anything visually for the user.

**Comment by carolinan** [comment]
Edit: Actually the verse can have a heading, the block has the "edit as HTML" option.

**Comment by youknowriad** [comment]
I think it should probably follow the same approach as other elements like "button". I think it relies on `color.button` in block.json in addition to it being supported at the root level. Right now the heading shows up too often IMO

**Comment by carolinan** [comment]
> Right now the heading shows up too often IMO

Agreed.


There is no Heading (element) block support equivalent to `buttonColor`. 
@scruffian implemented the `buttonColor` in https://github.com/WordPress/gutenberg/pull/41822 so he might have some idea on how to implement it for headings.

**Pull request merged**